README.md
# MSSQL-Database-Insights widget

This collection of widgets are designed to provide insights into MSSQL Database to further extend the built-in default widgets.

Where possible all of these widgets will include more detail when you click *_Show Details_* from the widget menu.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/show-detail.png" alt="Show Details" />

## Change Log

See the [Change Log](./src/CHANGELOG.md) for the full changes.

## Installation

The current release will be available through the Extensions Marketplace in Sql Ops Studio. When you click ***install*** in market place, you will be taken to the project's releases page. Locate the release you want and download the VSIX. Then install that via the File menu ***Install Extension from VSIX Package*** option in Sql Ops Studio.

Current and Pre-releases will be available from the [Releases](https://github.com/Matticusau/sqlops-mssql-db-insights/releases) tab of the projects repository. Simply download the VSIX of the release you want, and use the ***Install Extension from VSIX Package*** option in Sql Ops Studio.

## Supported SQL Server Versions

These widgets have been tested against the following SQL Server versions:

* SQL Server 2008R2
* SQL Server 2012
* SQL Server 2014
* SQL Server 2016
* SQL Server 2017 (Windows & linux)
* Azure SQL Db

If you find any issues using these widgets on these supported SQL Server versions, or any other versions please create an issue as we would like to make these available for as many releases as possible.

Thanks to the following people for testing:

* Sreekanth [@mssqltrek](https://twitter.com/mssqltrek)

***We are looking for testers to confirm other environments.*** So if you find they do work on other releases let me know, and credit will be given.

## Dashboard Tab

When the extension is loaded it will add a Dashboard tab. You can edit your workspace settings in the *dashboard.server.tabs* section to include this on your specific projects.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-insights-tab.png" alt="mssql-db-insights-tab.png" style="width:800px;" />

## mssql-db-spaceused

This Database Dashboard widget includes information on the current used space within a Database. Information will be shown in the form of a pie chart displaying the percentage of used space verses free space.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-spaceused.png" alt="mssql-db-spaceused.png" />

You can access more information about the space usaged in the detailed fly-out displayed when you select "..." on the widget.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-spaceused-details.png" alt="mssql-db-spaceused-details.png" />

To enable this widget add the following json to either your user or workspace settings in the *dashboard.database.widgets* section.

```json
{
    "widget": {
        "mssql-db-spaceused.insight": {
            "cacheId": "1d7cba8b-c87a-4bcc-ae54-2f40a5503a90"
        }
    }
}
```

## mssql-db-spaceused-filetype

This Database Dashboard widget includes information on the current used space within a Database broken down by the various file types (ROWS, LOG). Information will be shown in the form of a bar chart displaying the percentage of used space verses free space for each file type.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-spaceused-filetype.png" alt="mssql-db-spaceused-filetype.png" />

You can access more information about the space usaged in the detailed fly-out displayed when you select "..." on the widget.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-spaceused-details.png" alt="mssql-db-spaceused-details.png" />

To enable this widget add the following json to either your user or workspace settings in the *dashboard.database.widgets* section.

```json
{
    "widget": {
        "mssql-db-spaceused-filetype.insight": {
            "cacheId": "1d7cba8b-c87a-4bcc-ae54-2f40a5503a90"
        }
    }
}
```

## mssql-db-vlfs

This Database Dashboard widget includes information on the number of VLfs in the current database on the SQL Instance. Information will be shown in the form of a bar chart.

This insight is ***not*** supported on Azure SQL Db.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-vlfs.png" alt="mssql-db-vlfs.png" />

You can access more information about the vlfs in the detailed fly-out displayed when you select "..." on the widget.

<img src="https://raw.githubusercontent.com/Matticusau/sqlops-mssql-db-insights/master/images/mssql-db-vlfs-details.png" alt="mssql-db-vlfs-details.png" />

To enable this widget add the following json to either your user or workspace settings in the *dashboard.database.widgets* section.

```json
{
    "widget": {
        "mssql-db-vlfs.insight": {
            "cacheId": "1d7cba8b-c87a-4bcc-ae54-2f40a5503a90"
        }
    }
}
```

## Vector Graphics

The vector graphics used in this tool are sourced from the following authors

Icons made by [Freepik](http://www.freepik.com) sourced from [Flaticon](https://www.flaticon.com/) and is licensed by [CC 3.0 BY](http://creativecommons.org/licenses/by/3.0/)
Icons made by [Gregor Cresnar](https://www.flaticon.com/authors/gregor-cresnar) sourced from [Flaticon](https://www.flaticon.com/) and is licensed by [CC 3.0 BY](http://creativecommons.org/licenses/by/3.0/)

## License

This project is released under the [MIT License](./LICENSE)

## Contributors

* Matticusau [GitHub](https://github.com/Matticusau) | [twitter](https://twitter.com/matticusau)


CHANGELOG.md
# Change Log

All notable changes to the "mssql-db-insights" extension will be documented in this file. Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- New SVG icons for dashboard containers

## [0.2.2]

### Added

- Multiple containers/dashboards
- DB Session insight
- DB Resource Consumption insight
- Autorefreh for certain insights

### Changed

- New gallery logo
- Fixed license url in TSQL
- Friendly column names in charts
- Fixed icon colors in fly out window

## [0.2.0]

### Added

- VLFs widget

### Changed

- Minor documentation fixes for marketplace support

## [0.2.0-beta1]

### Added

- Migration of code from [https://github.com/Matticusau/sqlops-widgets](https://github.com/Matticusau/sqlops-widgets)

Build.ps1
[CmdLetBinding()]
Param (
    [Parameter()]
    [string]$BuildDir
)

[string]$scriptPath = $PSScriptRoot;

# set the location history 
Push-Location;

# make sure the releases folder exists
if ($null -eq $BuildDir -or $BuildDir.Length -eq 0)
{
    $BuildDir = Join-Path -Path $scriptPath -ChildPath 'build';
}
if (-not(Test-Path -Path $BuildDir))
{
    New-Item -Path $BuildDir -ItemType Directory -Force | Out-Null;
}

# Get the packages to build in this repo
$packageFiles = Get-ChildItem -Path $scriptPath -Recurse -Filter 'package.json'

if ($packageFiles.Count -eq 0)
{
    Write-Error -Message 'No Extensions found to package'
    Exit
}

# set our working location to the build folder
Set-Location -Path $BuildDir;

# process each package found
foreach ($package in $packageFiles)
{
    Write-Verbose -Message "Processing $($package.fullname)";

    # try
    # {
        Set-Location -Path $package.DirectoryName;
        vsce.cmd package;
        Get-ChildItem -Path $package.DirectoryName -Filter '*.vsix' | Move-Item -Destination $BuildDir -Force;
    # }
    # catch 
    # {
        
    # }

    Write-Verbose -Message ('Finished processing package {0}' -f $packageJson.name);

}

# return to the previous user location
Pop-Location;

Write-Verbose -Message 'Extension Packaging Complete';

src/utils.ts
//
// Author:  Matt Lavery
// Date:    02/07/2018
// Purpose: Utils
//
// When         Who         What
// ------------------------------------------------------------------------------------------
// 02/07/2018   MLavery     Strictly set 'any' types to fix src\extension.ts(50,55): error TS7006: Parameter 'connection' implicitly has an 'any' type.
//

/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the Source EULA. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

'use strict';

import * as fs from 'fs-extra';
import * as handlebars from 'handlebars';
import * as path from 'path';
import * as vscode from 'vscode';

import * as Constants from './constants';
import * as LocalizedConstants from './localizedConstants';

/**
 * Helper to log messages to the developer console if enabled
 * @param msg Message to log to the console
 */
export function logDebug(msg: any): void {
    let config = vscode.workspace.getConfiguration(Constants.extensionConfigSectionName);
    let logDebugInfo = config[Constants.configLogDebugInfo];
    if (logDebugInfo === true) {
        let currentTime = new Date().toLocaleTimeString();
        let outputMsg = '[' + currentTime + ']: ' + msg ? msg.toString() : '';
        console.log(outputMsg);
    }
}

export function renderTemplateHtml(extensionPath: string, templateName: string, templateValues: object): Thenable<string> {
    // let templatePath = path.join(extensionPath, 'resources', templateName);
    let templatePath: string = path.join(extensionPath, 'resources', templateName);

    // 1) Read the template from the disk
    // 2) Compile it as a handlebars template and render the HTML
    // 3) On failure, return a simple string as an error
    return fs.readFile(templatePath, 'utf-8')
        .then((templateText: string) => {
            let template = handlebars.compile(templateText);
            return template(templateValues);
        })
        .then(
            undefined,
            (error: any) => {
                logDebug(error);
                return LocalizedConstants.msgErrorLoadingTab;
            }
        );
}




BuildVSIX.ps1
#
# Author:  Matt Lavery
# Date:    03/07/2018
# Purpose: Build VSIX package for project
#
# When         Who         What
# ------------------------------------------------------------------------------------------
# 03/07/2018   MLavery     Updated for new project structure
#

[CmdLetBinding()]
Param (
    [Parameter()]
    [string]$BuildDir
)

[string]$scriptPath = $PSScriptRoot;

# set the location history 
Push-Location;

# make sure the releases folder exists
if ($null -eq $BuildDir -or $BuildDir.Length -eq 0)
{
    $BuildDir = Join-Path -Path $scriptPath -ChildPath '_vsix';
}
if (-not(Test-Path -Path $BuildDir))
{
    New-Item -Path $BuildDir -ItemType Directory -Force | Out-Null;
}

# Get the packages to build in this repo
$packageFiles = Get-ChildItem -Path $scriptPath -Recurse -Filter 'package.json'

if ($packageFiles.Count -eq 0)
{
    Write-Error -Message 'No Extensions found to package'
    Exit
}

# set our working location to the build folder
Set-Location -Path $BuildDir;

# process each package found
foreach ($package in $packageFiles)
{
    Write-Verbose -Message "Processing $($package.fullname)";

    # try
    # {
        Set-Location -Path $package.DirectoryName;
        vsce.cmd package;
        Get-ChildItem -Path $package.DirectoryName -Filter '*.vsix' | Move-Item -Destination $BuildDir -Force;
    # }
    # catch 
    # {
        
    # }

    Write-Verbose -Message ('Finished processing package {0}' -f $packageJson.name);

}

# return to the previous user location
Pop-Location;

Write-Verbose -Message 'Extension Packaging Complete';

src/controllers/mainController.ts
//
// Author:  Matt Lavery
// Date:    02/07/2018
// Purpose: Main Controller
//
// When         Who         What
// ------------------------------------------------------------------------------------------
// 02/07/2018   MLavery     Strictly set 'any' types to fix src\extension.ts(50,55): error TS7006: Parameter 'connection' implicitly has an 'any' type.
//

/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the Source EULA. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

'use strict';

import * as vscode from 'vscode';
import * as sqlops from 'sqlops';
import * as Utils from '../utils';
import ControllerBase from './controllerBase';
import * as fs from 'fs';
import * as path from 'path';
import * as dbprops from './dbPropsController';

/**
 * The main controller class that initializes the extension
 */
export default class MainController extends ControllerBase {
    // PUBLIC METHODS //////////////////////////////////////////////////////
    /**
     * Deactivates the extension
     */
    public deactivate(): void {
        Utils.logDebug('Main controller deactivated');
    }

    public activate(): Promise<boolean> {
        
        // const webviewExampleHtml = fs.readFileSync(path.join(__dirname, 'webviewExample.html')).toString();
        // const buttonHtml = fs.readFileSync(path.join(__dirname, 'button.html')).toString();
        // const counterHtml = fs.readFileSync(path.join(__dirname, 'counter.html')).toString();

        // let countWidget: sqlops.DashboardWebview;
        // let buttonWidget: sqlops.DashboardWebview;
        // let count = 0;

        // let dialog: sqlops.ModalDialog = sqlops.window.createDialog('Flyout extension');
        // dialog.html = '<div>This is a flyout extension.</div>';

        // sqlops.dashboard.registerWebviewProvider('webview-count', e => {
        //     e.html = counterHtml;
        //     countWidget = e;
        // });
        // sqlops.dashboard.registerWebviewProvider('webview-button', e => {
        //     e.html = buttonHtml;
        //     buttonWidget = e;
        //     e.onMessage(event => {
        //         if (event === 'openFlyout') {
        //             dialog.open();
        //         } else {
        //             count++;
        //             countWidget.postMessage(count);
        //         }
        //     });
        // });
        // sqlops.dashboard.registerWebviewProvider('webviewExample', e => {
        //     e.html = webviewExampleHtml;
        // });

        // const dbpropertiesHtml = fs.readFileSync(path.join(__dirname, 'dbProperties.html')).toString();
        // console.log('dbpropertiesHtml'+dbpropertiesHtml);
        
        sqlops.dashboard.registerWebviewProvider('dbproperties', (e: any) => {
            dbprops.getDbProperties()
            .then(dbpropertiesValues => {
                console.log('dbpropertiesValues:'+JSON.stringify(dbpropertiesValues));
                Utils.renderTemplateHtml(path.join(__dirname, '..'), 'dbProperties.html', dbpropertiesValues)
                    .then(html => {
                        console.log('dbpropertiesHtml'+html);
                        console.log('setting e.html');
                        e.html = html;
                    });
                });
            //console.log('setting e.html to dbpropertiesHtml');
            //e.html = dbpropertiesHtml;
        });

        //
        // register the commands
        //
        vscode.commands.registerCommand('extension.sayHello', () => {
            // The code you place here will be executed every time your command is executed
    
            // Display a message box to the user
            vscode.window.showInformationMessage('Hello World!');
        });

        vscode.commands.registerCommand('extension.showCurrentConnection', () => {
            // The code you place here will be executed every time your command is executed
    
            // Display a message box to the user
            sqlops.connection.getCurrentConnection().then((connection: any) => {
                let connectionId = connection ? connection.connectionId : 'No connection found!';
                vscode.window.showInformationMessage(connectionId);
            }, (error: any) => {
                 console.info(error);
            });
        });

        return Promise.resolve(true);
    }
}



src/controllers/dbPropsController.ts
//
// Author:  Matt Lavery
// Date:    03/07/2018
// Purpose: db Properites
//
// When         Who         What
// ------------------------------------------------------------------------------------------
// 03/07/2018   MLavery     Initial coding
//


'use strict';

import * as vscode from 'vscode';
import * as sqlops from 'sqlops';

export function getDbProperties(): Thenable<any> {
    
    return sqlops.connection.getCurrentConnection().then((connection: any) => {
        console.log('connection: '+JSON.stringify(connection));
        if (connection.connectionProvider === 'MSSQL') {
            let dbpropertiesValues = { name: 'Test DB', collation: 'Latin something something' };
            return dbpropertiesValues;
        }
        else
        {
            return { name: '<< Non MSSQL Database >>' };
        }
        // vscode.window.showInformationMessage(connectionId);
    }, (error: any) => {
         console.info(error);
         return {};
    });
}


src/extension.ts
//
// Author:  Matt Lavery
// Date:    02/07/2018
// Purpose: extension entry module
//
// When         Who         What
// ------------------------------------------------------------------------------------------
// 02/07/2018   MLavery     Strictly set 'any' types to fix src\extension.ts(50,55): error TS7006: Parameter 'connection' implicitly has an 'any' type.
//

'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

// The module 'sqlops' contains the SQL Operations Studio extensibility API
// This is a complementary set of APIs that add SQL / Data-specific functionality to the app
// Import the module and reference it with the alias sqlops in your code below

// import * as sqlops from 'sqlops';

// import the controller
import ControllerBase from './controllers/controllerBase';
import MainController from './controllers/mainController';

let controllers: ControllerBase[] = [];


// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

    console.log('Starting activation');

    // taken from the extensionSamples
    let activations: Promise<boolean>[] = [];

    // Start the main controller
    let mainController = new MainController(context);
    controllers.push(mainController);
    context.subscriptions.push(mainController);
    activations.push(mainController.activate());

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "mssql-db-insights" is now active!');

    // activations - from extensionSamples
    return Promise.all(activations)
        .then((results: boolean[]) => {
            for (let result of results) {
                if (!result) {
                    return false;
                }
            }
            return true;
        });
}

// this method is called when your extension is deactivated
export function deactivate() {
    // deactive any controllers we loaded
    for (let controller of controllers) {
        controller.deactivate();
    }
}